cdap-app-fabric/src/test/java/co/cask/cdap/app/runtime/monitor/RuntimeMonitorTest.java

```diff
@@ -0,0 +1,198 @@
+
+    Map<String, String> topics = new HashMap<>();
+    topics.put(Constants.AppFabric.PROGRAM_STATUS_RECORD_EVENT_TOPIC, "status");
+
+    httpServer.createContext("/v3/runtime/monitor/topics", new HttpHandler() {
+      public void handle(HttpExchange exchange) throws IOException {
+        byte[] response = GSON.toJson(topics).getBytes();
+        exchange.sendResponseHeaders(HttpURLConnection.HTTP_OK, response.length);
+        exchange.getResponseBody().write(response);
+        exchange.close();
+      }
+    });
+
+    ConnectionConfig connectionConfig = ConnectionConfig.builder()
+      .setHostname("127.0.0.1")
+      .setPort(8000)
+      .setSSLEnabled(false)
+      .build();
+    ClientConfig.Builder clientConfigBuilder = ClientConfig.builder()
+      .setDefaultReadTimeout(15000)
+      .setConnectionConfig(connectionConfig);
+
+    int limit = 2;
+    MessagingContext messagingContext = new MultiThreadMessagingContext(messagingService);
+    RuntimeMonitor runtimeMonitor = new RuntimeMonitor(new ProgramRunId("test", "app1", ProgramType.WORKFLOW, "p1",
+                                                                        "run1"),
+                                                       cConf, messagingContext.getMessagePublisher(),
+                                                       clientConfigBuilder.build());
+
+    Map<String, List<MonitorMessage>> messages = new LinkedHashMap<>();
+    ArrayList<MonitorMessage> list = new ArrayList<>();
+    list.add(new MonitorMessage("1", "message1"));
+    list.add(new MonitorMessage("2", "message2"));
+    list.add(new MonitorMessage("3", "message3"));
+    list.add(new MonitorMessage("4", "message4"));
+    list.add(new MonitorMessage("5", "message5"));
+    list.add(new MonitorMessage("6", "message6"));
+    list.add(new MonitorMessage("7", "message7"));
+    list.add(new MonitorMessage("8", "message8"));
+    list.add(new MonitorMessage("9", "message9"));
+    list.add(new MonitorMessage("10", "message10"));
+
+    messages.put("status", list);
+
+    httpServer.createContext("/v3/runtime/metadata", new HttpHandler() {
+      int count = 0;
+
+      public void handle(HttpExchange exchange) throws IOException {
+        Map<String, List<MonitorMessage>> toSend = new LinkedHashMap<>();
+        ArrayList<MonitorMessage> list = new ArrayList<>();
+        int start = count;
+        int i = 0;
+        for (MonitorMessage message : messages.get("status")) {
+          if (start <= i && i < start + limit) {
+            list.add(message);
+            count++;
+          }
+          i++;
+        }
+        toSend.put("status", list);
+
+        byte[] response = GSON.toJson(toSend).getBytes();
+        exchange.sendResponseHeaders(HttpURLConnection.HTTP_OK, response.length);
+        exchange.getResponseBody().write(response);
+        exchange.close();
+
+      }
+    });
+
+    runtimeMonitor.startAndWait();
+    Thread.sleep(5000);
```

Comment:
Try not to sleep in unit-test. This will make test runs unnecessarily long. Instead, you should keep polling from TMS until you see all expected messages (with a small sleep between poll, say, 100ms).